### ISSUE TITLE
Implement a typeclass hierarchy for Accelerate

### ISSUE THREAD

**Issue by NickSeagull** [AccelerateHS/accelerate#397]
It would be very nice to have a type class hierarchy for the `Acc (Array sh a)` types.

Working with standard category theory abstractions is very useful. To be able to work thinking in Monoids, Functors, etc... makes stuff easier most of the time.

The problem is that we have a constraint (`Elt a`) that we have to satisfy, so we cannot instantiate the standard type classes from `Prelude`, speaking of which, we mostly redefined here. It is not crazy to redefine the type class hierarchy here too.

As a first approach I thought about using the [constrained-categories](http://hackage.haskell.org/package/constrained-categories) package, which allows us to define our own "constrained world" (constrained category). For accelerate we could write something like this:

```haskell
-- Hiding Acc here might not be good as tmcdonell noted
type AccArray sh a = Acc (Array sh a)

-- Definition of the morphism (how functions work) using GADTs
data Accel a b where
  Accel :: (Elt a, Elt b) => (Exp a -> Exp b) -> Accel a b

-- How our world behaves
instance Category (Accel a b) where
  type Object (Accel a b) a = Elt a       -- Here we define the constraint for our objects
  id = Accel id                         -- The identity function
  Accel f . Accel g = Accel ( f . g )   -- Function composition
```
_Note: I'm writing this of the top of my head, it might not compile_

So now we can instantiate type classes defined in the `constrained-categories` package:

```haskell
instance Shape sh => Functor (AccArray sh) Accel Accel where
  fmap = constrainedFmap map            -- We are using the Accelerate map here

-- We can also define our own that are not defined there:
class Category t => Semigroup m t where           -- Semigroup in the world of 't'
  append :: (Object t a) => m a -> m a -> m a     -- 'a' is an object of 't'
  
instance (Shape sh) => Semigroup (AccArray sh) Accel where
  append = (++)
```

In the functor instance we are saying `Accel` twice. Because conceptually Functors are a pathway from one category to another, we define the `from` and the `destination` category. **We do not really need this**, as we always will stay within Accelerate.

Maybe we should just define the [purescript type class hierarchy](http://blog.functorial.com/posts/2015-12-06-Counterexamples.html)? It is very complete and most of the people like it.

**Comment by tmcdonell** [comment]
I think this makes more sense at the `Exp` level, actually, because you can define new instances of `Elt` but not so much for `Arrays`. Having `Functor` on `Exp` might have made some `linear-accelerate` easier to implement, actually, so experimenting there with what works might be illuminating.

However, that seems to be about as far as we can go. While we could implement `Functor`, since Accelerate is first order (with some built-in second-order combinators), we couldn't implement `Apply` (/`Applicative`). So, pulling in the whole of constrained-categories seems like overkill for just one thing, but I could be missing something...

**Comment by tmcdonell** [comment]
I think a while back somebody was interested in using [subhask](http://hackage.haskell.org/package/subhask) as the base class hierarchy, but that may be only tangentially related...

**Comment by NickSeagull** [comment]
Yes, I was thinking about the fact that we cannot go with `Apply/Applicative` and further. But still we could define `Traversable`. `Foldable`, `Semigroup` and `Monoid`, right?

**Comment by tmcdonell** [comment]
It looks like all the `Traversable` functions require `Applicative` or `Monad`. The others seem to make sense though.

I don't know if it is possible to define these for both `Exp` and `Acc` types while keeping a similar interface to the standard definitions (e.g. without a multi-param type class so that we can explicitly mention which level we are at; we may not be able to get around that though.)

**Comment by tmcdonell** [comment]
Related question. I implemented the `Num` class hierarchy pretty much directly as it exists in the Haskell prelude, which seemed to be the path of least surprise. But, that is known to be unsatisfactory in various ways, so perhaps we should instead pick a different hierarchy?

**Comment by NickSeagull** [comment]
Could you elaborate more on the "different hierarchy" subject? Which one would we pick @tmcdonell ?

**Comment by tmcdonell** [comment]
I'm not sure what the best model would be; open to suggestions.

`SubHask` has a very detailed hierarchy, for example, but maybe that is too much? (That is, would it be too unrecognisable for a person familiar with Haskell trying to pick up Accelerate and create too much additional friction?)

**Comment by tmcdonell** [comment]
c.f. #169

**Comment by tmcdonell** [comment]
I've added some `Functor` and `Semigroup` shenanigans for `Exp` terms. I'm still not sure how or why we should add this at the `Acc` level though; I'd be interested in an example use case, otherwise I may just call this done...

**Comment by NickSeagull** [comment]
I think that it makes sense to have it only at the `Exp` level, as I think that it doesnt make too much sense to have it for an operation abstraction.

**Comment by tmcdonell** [comment]
Sounds good to me. Once I fix the build for 7.10 I will merge this into master (:

If we find a good use case for lifting this to the `Acc` level we can revisit this decision.
